wgpu/src/backend/webgpu.rs

```diff
@@ -1417,83 +1523,176 @@ impl crate::context::Context for ContextWebGpu {
+                        WebShaderCompilationInfo::Wgsl {
+                            source: code.to_string(),
+                        },
+                    ))
+                }
+                #[cfg(feature = "naga-ir")]
+                crate::ShaderSource::Naga(ref module) => {
+                    validate_transformed_shader_module(module, "", desc).map(|v| {
+                        (
+                            v,
+                            WebShaderCompilationInfo::Transformed {
+                                compilation_info: CompilationInfo { messages: vec![] },
+                            },
+                        )
+                    })
+                }
+                crate::ShaderSource::Dummy(_) => {
+                    panic!("found `ShaderSource::Dummy`")
+                }
             }
-            #[cfg(feature = "wgsl")]
-            crate::ShaderSource::Wgsl(ref code) => webgpu_sys::GpuShaderModuleDescriptor::new(code),
-            #[cfg(feature = "naga-ir")]
-            crate::ShaderSource::Naga(module) => {
-                use naga::{back, valid};
-
-                let mut validator = valid::Validator::new(
-                    valid::ValidationFlags::all(),
-                    valid::Capabilities::all(),
-                );
-                let module_info = validator.validate(&module).unwrap();
+        }
 
-                let writer_flags = naga::back::wgsl::WriterFlags::empty();
-                let wgsl_text =
-                    back::wgsl::write_string(&module, &module_info, writer_flags).unwrap();
-                webgpu_sys::GpuShaderModuleDescriptor::new(wgsl_text.as_str())
-            }
-            crate::ShaderSource::Dummy(_) => {
-                panic!("found `ShaderSource::Dummy`")
-            }
+        #[cfg(any(feature = "spirv", feature = "glsl", feature = "naga-ir"))]
+        fn validate_transformed_shader_module(
+            module: &naga::Module,
+            source: &str,
+            desc: &crate::ShaderModuleDescriptor<'_>,
+        ) -> Result<webgpu_sys::GpuShaderModuleDescriptor, crate::CompilationInfo> {
+            use naga::{back, valid};
+            let mut validator =
+                valid::Validator::new(valid::ValidationFlags::all(), valid::Capabilities::all());
+            let module_info = validator
+                .validate(module)
+                .map_err(|err| wgc::pipeline::ShaderError {
+                    source: source.to_string(),
+                    label: desc.label.map(|s| s.to_string()),
+                    inner: Box::new(err),
+                })
+                .map_err(|err| crate::CompilationInfo {
+                    messages: vec![crate::CompilationMessage {
+                        message: err.to_string(),
+                        message_type: crate::CompilationMessageType::Error,
+                        span: err
+                            .inner
+                            .spans()
```

Comment:
as before, why merge the errors and then take arbitrarily the first span instead of exposing all errors als separate messages

EDIT: ah I think this is different here because this is a `WithSpan<ValidationError>` if I'm not mistaken. Hmm still strange though to pick the first span and ignore the rest :/. Maybe nothing to be done about it right now.

Reply:
Yeah, I'm not certain what the best solution here is. I could
- do nothing and keep it as is
- add a Vec<> for the compilation spans to the `CompilationMessage` type, deviating from the standard
- find a few good examples and open an issue in the repository where the work on the WebGPU specification happens (surely there are cases where an error clearly has multiple spans, right?)
- report multiple spans as multiple errors? But that seems wrong
- join the spans? Also seems wrong

Reply:
The other spans should be in the error message, which is hopefully good enough.